### PR TITLE
Gas based keeper rewards

### PR DESCRIPTION
**Pull request by leomassazza**
The objective of this PR is to get the cost of a transaction based on gas price. It is used to dynamically calculate the keeper rewards for settlement, flag and liquidate, and to calculate the required margin to pay the keeper upon liquidation.

It  makes use of a node that gets a transaction cost based on current gas price, and then the keeper rewards are added to the cost. 

---
Detailed TODO list:
- [x] Add mock TxGasPriceOracle
- [x] Add node for keeper rewards gas cost  
- [x] Add KeeperCost library to storage that calls the node
- [x] Add node to bootstrap
- [x] Configure node with all zeros (not alter current tests)
- [x] Tests pass with node
- [x] Test not configurable by non-owner
- [x] Test to check revert if not configured
- [x] Add cost of execution on settlement
- [x] Add cost of execution on flag
- [x] Add cost of execution on liquidation
- [x] Add cost of flag/liquidate to isElegibleForLiquidation
- [x] Dynamic min and max keeper reward caps
- [x] Test settlement cost is used on settlement
- [x] Test caps on flag + liquidate
- [x] Test flag / liquidate for 1 margin 1 small position 
- [x] Test flag / liquidate for 1 margin 2 small positions (multiple feeds for positions) 
- [x] Test flag / liquidate for 1 margin 1 large position (multiple steps)
- [x] Test flag / liquidate for 2 margins 1 small position (multiple feeds for margin)
- [x] Test flag/liquidate cost is added to isElegibleForLiquidation
- [x] Test like Order.marginValidation.test with caps
- [x] Add extra reward for liquidators (execution reward on top of cost)
- [x] On tests, confirm liquidator got the rewards
- [x] review comments and docs and add missing ones 
- [x] cleanup (no console logs or dev stuff)
- [x] lint
- [ ] all test pass green

### CODE REVIEW

File: markets/perps-market/contracts/storage/KeeperCosts.sol [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+            KIND_LIQUIDATION_ELIGIBILITY
+        );
+
+        sUSDCost = INodeModule(factory.oracle)
+            .processWithRuntime(self.feedId, runtimeKeys, runtimeValues)
+            .price
+            .toUint();
+    }
+
+    function getRuntime(
```
Comment (sunnyvempati) [inline markets/perps-market/contracts/storage/KeeperCosts.sol:121]:
```suggestion
    function _processWithRuntime(
```

seems like everytime `getRuntime` is called, we usually call process so can just move it all into this private method

File: markets/perps-market/contracts/storage/AsyncOrder.sol [hunk 2]
```diff
@@ -540,8 +547,9 @@ library AsyncOrder {
         // do same thing for liquidation rewards and account for minimum liquidation margin
         uint256 requiredLiquidationRewardMargin = GlobalPerpsMarketConfiguration
             .load()
-            .minimumLiquidationReward(
-                currentTotalLiquidationRewards + newLiquidationReward - oldLiquidationReward
+            .minimumKeeperReward(
+                currentTotalLiquidationRewards + newLiquidationReward - oldLiquidationReward,
```
Comment (sunnyvempati) [inline markets/perps-market/contracts/storage/AsyncOrder.sol:551]: the `costOfExecution` here is based on the old position and needs to be recalculated for new position. and added to this line

```
uint256 newTotalLiquidationRewards = currentTotalLiquidationRewards + newLiquidationReward - oldLiquidationReward + newCostOfExecution;
uint256 requiredLiquidationRewardMargin = GlobalPerpsMarketConfiguration
  .load().minimumKeeperReward(newTotalLiquidationRewards, newCostOfExecution);
```

**Review by sunnyvempati — CHANGES_REQUESTED** [review]
hope the comments make sense.  hmu in the AM with any q's.  we're very close!  great work on this

File: markets/perps-market/contracts/modules/LiquidationModule.sol [hunk 3]
```diff
@@ -41,16 +43,21 @@ contract LiquidationModule is ILiquidationModule, IMarketEvents {
...
+            (bool isEligible, , , , , , ) = account.isEligibleForLiquidation();
 
             if (isEligible) {
-                account.flagForLiquidation();
-                liquidationReward = _liquidateAccount(account);
+                uint flagCost = account.flagForLiquidation();
+                liquidationReward = _liquidateAccount(
+                    account,
+                    flagCost,
+                    account.getTotalCollateralValue(),
```
[the remark this answers is not in the document]
Comment (leomassazza) [reply markets/perps-market/contracts/modules/LiquidationModule.sol:53]: Thanks for noticing... you discovered a hidden bug here. It shouldn't be zero.

File: markets/perps-market/contracts/modules/LiquidationModule.sol [hunk 4]
```diff
@@ -41,16 +43,21 @@ contract LiquidationModule is ILiquidationModule, IMarketEvents {
...
-            (bool isEligible, , , , , ) = account.isEligibleForLiquidation();
+            (bool isEligible, , , , , , ) = account.isEligibleForLiquidation();
 
             if (isEligible) {
-                account.flagForLiquidation();
-                liquidationReward = _liquidateAccount(account);
+                uint flagCost = account.flagForLiquidation();
+                liquidationReward = _liquidateAccount(
+                    account,
+                    flagCost,
```
[the remark this answers is not in the document]
Comment (leomassazza) [reply markets/perps-market/contracts/modules/LiquidationModule.sol]: depending on the config flagCost can be zero even if there was a flag here. (almost on some tests I got this config)

**Review by sunnyvempati — APPROVED** [review]
couple changes; pre-approving.  awesome work on this

**Comment by noahlitvin** [comment]
Closing in favor of #1904

**Pull request closed without merging**